IMPORT THE TEST AND TRAIN DATA

In [1]:
import pandas as pd
df_Train=pd.read_csv("Final_Train.csv")
print (df_Train)

                       Qualification           Experience Rating  \
0              BHMS, MD - Homeopathy  24 years experience   100%   
1       BAMS, MD - Ayurveda Medicine  12 years experience    98%   
2     MBBS, MS - Otorhinolaryngology   9 years experience    NaN   
3                BSc - Zoology, BAMS  12 years experience    NaN   
4                               BAMS  20 years experience   100%   
...                              ...                  ...    ...   
5956                  MBBS, MS - ENT  19 years experience    98%   
5957                            MBBS  33 years experience    NaN   
5958                            MBBS  41 years experience    97%   
5959     MBBS, MD - General Medicine  15 years experience    90%   
5960         BDS, MDS - Orthodontics  17 years experience   100%   

                              Place           Profile  \
0               Kakkanad, Ernakulam         Homeopath   
1             Whitefield, Bangalore          Ayurveda   
2        Mat

In [2]:
import pandas as pd
df_Test=pd.read_csv("Final_test.csv")
print (df_Test)

                                        Qualification           Experience  \
0                                                MBBS  35 years experience   
1          MBBS, Diploma in Otorhinolaryngology (DLO)  31 years experience   
2                                          MBBS, DDVL  40 years experience   
3                                                BAMS   0 years experience   
4     BDS, MDS - Conservative Dentistry & Endodontics  16 years experience   
...                                               ...                  ...   
1982                         BAMS, Doctor of Medicine  10 years experience   
1983   MD - Dermatology , Venereology & Leprosy, MBBS   4 years experience   
1984         MBBS, MD / MS - Obstetrics & Gynaecology   9 years experience   
1985                                              BDS  12 years experience   
1986   MBBS, MD - Dermatology , Venereology & Leprosy   8 years experience   

     Rating                        Place           Profile  \
0

COMBINING TEST AND TRAIN DATA

In [4]:
df_Final=df_Train.append(df_Test,sort=False)

In [5]:
df_Final.head(10)

Qualification           Experience Rating  \
0           BHMS, MD - Homeopathy  24 years experience   100%   
1    BAMS, MD - Ayurveda Medicine  12 years experience    98%   
2  MBBS, MS - Otorhinolaryngology   9 years experience    NaN   
3             BSc - Zoology, BAMS  12 years experience    NaN   
4                            BAMS  20 years experience   100%   
5                            BAMS   8 years experience    NaN   
6                            BHMS  42 years experience    NaN   
7                             BDS  10 years experience    99%   
8     MBBS, MD - General Medicine  14 years experience    NaN   
9                        BSc, BDS  23 years experience    NaN   

                          Place           Profile  \
0           Kakkanad, Ernakulam         Homeopath   
1         Whitefield, Bangalore          Ayurveda   
2    Mathikere - BEL, Bangalore    ENT Specialist   
3  Bannerghatta Road, Bangalore          Ayurveda   
4         Keelkattalai, Chennai          Ayurveda   
5                Porur, Chennai          Ayurveda   
6             Karol Bagh, Delhi         Homeopath   
7            Arekere, Bangalore           Dentist   
8           Old City, Hyderabad  General Medicine   
9             Athani, Ernakulam           Dentist   

                                  Miscellaneous_Info   Fees  
0               100% 16 Feedback Kakkanad, Ernakulam  100.0  
1              98% 76 Feedback Whitefield, Bangalore  350.0  
2                                                NaN  300.0  
3  Bannerghatta Road, Bangalore ?250 Available on...  250.0  
4              100% 4 Feedback Keelkattalai, Chennai  250.0  
5                                                NaN  100.0  
6                                                NaN  200.0  
7  Dental Fillings Crowns and Bridges Fixing Impa...  200.0  
8                                                NaN  100.0  
9                                                NaN  100.0

EXTRACTING QUALIFICATION COLUMNS

In [6]:
import numpy as np
x=np.where(df_Final.Qualification.str.contains('(MBBS|BDS|MD|MS|LCEH|DDVL|BAMS|BSAM|BHMS|DHMS|DNB|Get inspired)')==False)

C:\Users\Asus\anaconda3\lib\site-packages\pandas\core\strings.py:1952: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


In [7]:
for i in x:
    print(df_Final.iloc[i,:].Qualification)

104     PhD - Orthodontics & Dentofacial Orthopaedics,...
224                                                  GCEH
753                                   39 years experience
829     Diplomate of National Board Exam, Diploma in O...
1030                 Diploma in Otorhinolaryngology (DLO)
1064              GCEH, Certificate in Child Health (CCH)
1265                               Diploma in Dermatology
1297                                              G.A.M.S
2005                                              G.A.M.S
2035    Fellowship in Oral implantlogy, Certified Adva...
2975                                           M.D. (E.H)
4503                                                  DDV
4757    Diploma in Otorhinolaryngology (DLO), Diploma ...
5182                                              G.A.M.S
443                                FRCS - General Surgery
471                                            FWFO, FPFA
1139                               Diploma in Dermatology
1159          

In [8]:
df_Final['MBBS']=df_Final.Qualification.str.extract('(MBBS)')
df_Final['BDS']=df_Final.Qualification.str.extract('(BDS)')
df_Final['MD']=df_Final.Qualification.str.extract('(MD)')
df_Final['MS']=df_Final.Qualification.str.extract('(MS)')
df_Final['LCEH']=df_Final.Qualification.str.extract('(LCEH)')
df_Final['DDVL']=df_Final.Qualification.str.extract('(DDVL)')
df_Final['BAMS']=df_Final.Qualification.str.extract('(BAMS)')
df_Final['BSAM']=df_Final.Qualification.str.extract('(BSAM)')
df_Final['BHMS']=df_Final.Qualification.str.extract('(BHMS)')
df_Final['DHMS']=df_Final.Qualification.str.extract('(DHMS)')
df_Final['DNB']=df_Final.Qualification.str.extract('(DNB)')
df_Final['Get']=df_Final.Qualification.str.extract('(Get inspired)')

CHECKING NULL VALUES

In [9]:
df_Final.isnull().sum()

Qualification            0
Experience               0
Rating                4392
Place                   31
Profile                  0
Miscellaneous_Info    3454
Fees                  1987
MBBS                  4164
BDS                   6158
MD                    4875
MS                    5067
LCEH                  7926
DDVL                  7720
BAMS                  6943
BSAM                  7938
BHMS                  6952
DHMS                  7883
DNB                   7386
Get                   7918
dtype: int64

In [10]:
df_Final['Rating'].fillna('0%',inplace = True)

In [11]:
df_Final['Rating'] = df_Final['Rating'].str.slice(stop=-1).astype(int)

DROPPING QUALIFICATION AND MISSLENEOUS INFO COLUMN

In [12]:
df_Final=df_Final.drop(['Qualification'],axis=1)
df_Final=df_Final.drop(['Miscellaneous_Info'],axis=1)

In [13]:
df_Final['Place']=df_Final['Place'].fillna('no city')
df_Final['Fees'].fillna((df_Final['Fees'].mean()),inplace=True)
df_Final['MBBS']=df_Final['MBBS'].fillna('none')
df_Final['BDS']=df_Final['BDS'].fillna('none')
df_Final['MD']=df_Final['MD'].fillna('none')

df_Final['MS']=df_Final['MS'].fillna('none')

df_Final['LCEH']=df_Final['LCEH'].fillna('none')
df_Final['DDVL']=df_Final['DDVL'].fillna('none')
df_Final['BAMS']=df_Final['BAMS'].fillna('none')
df_Final['BSAM']=df_Final['BSAM'].fillna('none')
df_Final['BHMS']=df_Final['BHMS'].fillna('none')

df_Final['DHMS']=df_Final['DHMS'].fillna('none')
df_Final['DNB']=df_Final['DNB'].fillna('none')

df_Final['Get']=df_Final['Get'].fillna('none')

In [14]:
df_Final.isnull().sum()

Experience    0
Rating        0
Place         0
Profile       0
Fees          0
MBBS          0
BDS           0
MD            0
MS            0
LCEH          0
DDVL          0
BAMS          0
BSAM          0
BHMS          0
DHMS          0
DNB           0
Get           0
dtype: int64

In [15]:
df_Final.columns

Index(['Experience', 'Rating', 'Place', 'Profile', 'Fees', 'MBBS', 'BDS', 'MD',
       'MS', 'LCEH', 'DDVL', 'BAMS', 'BSAM', 'BHMS', 'DHMS', 'DNB', 'Get'],
      dtype='object')

In [16]:
df_Final.head()

Experience  Rating                         Place         Profile  \
0  24 years experience     100           Kakkanad, Ernakulam       Homeopath   
1  12 years experience      98         Whitefield, Bangalore        Ayurveda   
2   9 years experience       0    Mathikere - BEL, Bangalore  ENT Specialist   
3  12 years experience       0  Bannerghatta Road, Bangalore        Ayurveda   
4  20 years experience     100         Keelkattalai, Chennai        Ayurveda   

    Fees  MBBS   BDS    MD  MS  LCEH  DDVL  BAMS  BSAM  BHMS  DHMS   DNB   Get  
0  100.0  none  none    MD  MS  none  none  none  none  BHMS  none  none  none  
1  350.0  none  none    MD  MS  none  none  BAMS  none  none  none  none  none  
2  300.0  MBBS  none  none  MS  none  none  none  none  none  none  none  none  
3  250.0  none  none  none  MS  none  none  BAMS  none  none  none  none  none  
4  250.0  none  none  none  MS  none  none  BAMS  none  none  none  none  none

PERFORMING SOME FEATURE ENGINEERING

In [17]:
df_Final['Profile']=df_Final['Profile'].astype('category')

In [18]:
df_Final['Experience1']=df_Final['Experience'].str.split('years experience').str[0]
df_Final['Experience2']=df_Final['Experience'].str.split('years experience').str[1]

In [19]:
df_Final=df_Final.drop(['Experience'],axis=1)
df_Final=df_Final.drop(['Experience2'],axis=1)

In [20]:
df_Final['Experience1']=df_Final['Experience1'].astype(int)

In [21]:
df_Final['Place1']=df_Final['Place'].str.split(',').str[0]
df_Final['Place2']=df_Final['Place'].str.split(',').str[1]

In [22]:
df_Final=df_Final.drop(['Place'],axis=1)

CHECKING DATA TYPES

In [23]:
df_Final.dtypes

Rating            int32
Profile        category
Fees            float64
MBBS             object
BDS              object
MD               object
MS               object
LCEH             object
DDVL             object
BAMS             object
BSAM             object
BHMS             object
DHMS             object
DNB              object
Get              object
Experience1       int32
Place1           object
Place2           object
dtype: object

In [24]:
from sklearn.preprocessing import LabelEncoder
LE=LabelEncoder()
df_Final["Profile"]=LE.fit_transform(df_Final['Profile'])
df_Final["MBBS"]=LE.fit_transform(df_Final['MBBS'])
df_Final["Place2"]=LE.fit_transform(df_Final['Place2'])
df_Final["BDS"]=LE.fit_transform(df_Final['BDS'])
df_Final["MD"]=LE.fit_transform(df_Final['MD'])
df_Final["Get"]=LE.fit_transform(df_Final['Get'])
df_Final["MS"]=LE.fit_transform(df_Final['MS'])
df_Final["LCEH"]=LE.fit_transform(df_Final['LCEH'])
df_Final["DDVL"]=LE.fit_transform(df_Final['DDVL'])
df_Final["BAMS"]=LE.fit_transform(df_Final['BAMS'])
df_Final["BSAM"]=LE.fit_transform(df_Final['BSAM'])
df_Final["BHMS"]=LE.fit_transform(df_Final['BHMS'])
df_Final["DHMS"]=LE.fit_transform(df_Final['DHMS'])
df_Final["DNB"]=LE.fit_transform(df_Final['DNB'])
df_Final["Place1"]=LE.fit_transform(df_Final['Place1'])

In [25]:
df_Final.dtypes

Rating           int32
Profile          int32
Fees           float64
MBBS             int32
BDS              int32
MD               int32
MS               int32
LCEH             int32
DDVL             int32
BAMS             int32
BSAM             int32
BHMS             int32
DHMS             int32
DNB              int32
Get              int32
Experience1      int32
Place1           int32
Place2           int32
dtype: object

SCALING THE DATA

In [26]:
from sklearn.preprocessing import MinMaxScaler
mms=MinMaxScaler()
scaled = mms.fit_transform(df_Final)
print(scaled)

[[1.         1.         0.1005291  ... 0.36363636 0.34578442 0.44444444]
 [0.98       0.         0.36507937 ... 0.18181818 0.98719317 0.        ]
 [0.         0.6        0.31216931 ... 0.13636364 0.52721451 0.        ]
 ...
 [0.         0.8        0.32057634 ... 0.13636364 0.7022412  0.11111111]
 [0.98       0.2        0.32057634 ... 0.18181818 0.03628602 0.        ]
 [0.         0.4        0.32057634 ... 0.12121212 0.64781217 0.11111111]]


SPLITTING INTO X AND Y

In [29]:
df_train=df_Final[0:5961]
df_test=df_Final[5961:]

In [30]:
x=df_train.drop(["Fees"],axis=1)
y=df_train["Fees"]

In [31]:
x.head()

Rating  Profile  MBBS  BDS  MD  MS  LCEH  DDVL  BAMS  BSAM  BHMS  DHMS  \
0     100        5     1    1   0   0     1     1     1     1     0     1   
1      98        0     1    1   0   0     1     1     0     1     1     1   
2       0        3     0    1   1   0     1     1     1     1     1     1   
3       0        0     1    1   1   0     1     1     0     1     1     1   
4     100        0     1    1   1   0     1     1     0     1     1     1   

   DNB  Get  Experience1  Place1  Place2  
0    1    1           24     324       4  
1    1    1           12     925       0  
2    1    1            9     494       0  
3    1    1           12      74       0  
4    1    1           20     367       1

In [32]:
y.head()

0    100.0
1    350.0
2    300.0
3    250.0
4    250.0
Name: Fees, dtype: float64

CHOOSING BEST RANDOM STATE BUILDING MODELS

In [33]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

for i in range(0,100):
    x_train, x_test,y_train,y_test = train_test_split(x, y, test_size=0.33, random_state=i)
    lr=LinearRegression()
    lr.fit(x_train,y_train)
    pred_train=lr.predict(x_train)
    pred_test=lr.predict(x_test)
    print(f"at random state {i} the training accuracy is:-{r2_score (y_train,pred_train)}")
    print(f"at random state {i} the testing accuracy is:-{r2_score (y_test,pred_test)}")
    print("\n")

at random state 0 the training accuracy is:-0.15264285778462594
at random state 0 the testing accuracy is:-0.15367555524779586


at random state 1 the training accuracy is:-0.15056990977882545
at random state 1 the testing accuracy is:-0.16156386356014096


at random state 2 the training accuracy is:-0.15813242004226546
at random state 2 the testing accuracy is:-0.14563704000468503


at random state 3 the training accuracy is:-0.16054803342972923
at random state 3 the testing accuracy is:-0.13529126532818592


at random state 4 the training accuracy is:-0.16429877406258775
at random state 4 the testing accuracy is:-0.1335716267559237


at random state 5 the training accuracy is:-0.15661314598662412
at random state 5 the testing accuracy is:-0.1474086719051202


at random state 6 the training accuracy is:-0.15466968190353092
at random state 6 the testing accuracy is:-0.1512735418711162


at random state 7 the training accuracy is:-0.17256607567425986
at random state 7 the testing accura

In [34]:
x_train, x_test,y_train,y_test = train_test_split(x, y, test_size=0.33, random_state=56)

In [35]:
lr.fit(x_train,y_train)

LinearRegression()

In [36]:
pred_test=lr.predict(x_test)

In [37]:
print(r2_score (y_test,pred_test))

0.15787269554161454


LASSO REGULARIZATION

In [38]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score

from sklearn.linear_model import Lasso
parameters = {'alpha':[.0001,.001,.01,.1,1,10],'random_state':(range(0,10))}
ls= Lasso()
clf = GridSearchCV(ls,parameters)
clf.fit(x_train,y_train)
print(clf.best_params_)

{'alpha': 0.0001, 'random_state': 0}


In [39]:
ls = Lasso(alpha=0.0001,random_state=0)
ls.fit(x_train,y_train)
ls.score(x_train,y_train)
pred_ls = ls.predict(x_test)
lss = r2_score(y_test,pred_ls)
lss

0.15787295823913594

RANDOM FOREST

In [40]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestRegressor
parameters = {'criterion':['mse','mae'],'max_features':["auto","sqrt","log2"]}
rf = RandomForestRegressor()
clf = GridSearchCV(rf,parameters)
clf.fit(x_train,y_train)
print(clf.best_params_)

{'criterion': 'mae', 'max_features': 'sqrt'}


In [41]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import cross_val_score
from sklearn.metrics import r2_score
rf = RandomForestRegressor(criterion="mae",max_features="sqrt")
rf.fit(x_train,y_train)
rf.score(x_train,y_train)
pred_rf = rf.predict(x_test)
rfs = r2_score(y_test,pred_rf)
print('R2 Score:', rfs*100)
rfscore = cross_val_score(rf,x,y,cv=5)
rfc = rfscore.mean()
print('Cross Val Score:',rfc*100)

R2 Score: 19.506689593503367
Cross Val Score: 17.78229530651629


SVR

In [42]:
from sklearn.svm import SVR
from sklearn.model_selection import cross_val_score
from sklearn.metrics import r2_score
sv = SVR()
sv.fit(x_train,y_train)
sv.score(x_train,y_train)
pred_rf = rf.predict(x_test)
svs = r2_score(y_test,pred_rf)
print('R2 Score:', svs*100)
svscore = cross_val_score(sv,x,y,cv=5)
svc = svscore.mean()
print('Cross Val Score:',rfc*100)

R2 Score: 19.506689593503367
Cross Val Score: 17.78229530651629


we are getting R2score value 19% in both SVR and RANDOM FOREST MODEL